README.org
#+TITLE: QuarkSuite Core (v2.1.0)

* Table of Contents :TOC@4:
- [[#introduction][Introduction]]
- [[#features][Features]]
- [[#packages][Packages]]
- [[#getting-started][Getting Started]]
  - [[#live][Live]]
  - [[#local-development][Local Development]]
    - [[#browserdeno][Browser/Deno]]
    - [[#node][Node]]
  - [[#offline][Offline]]
- [[#handbook][Handbook]]
- [[#api][API]]
- [[#demos][Demos]]
- [[#contributing][Contributing]]
- [[#concept][Concept]]
- [[#project-objectives][Project Objectives]]
  - [[#availability][Availability]]
  - [[#adaptability][Adaptability]]
  - [[#invisibility][Invisibility]]
  - [[#interoperability][Interoperability]]
  - [[#stability][Stability]]
- [[#credit][Credit]]

* Introduction

QuarkSuite Core is the engine of the [[https://github.com/quarksuite][emerging QuarkSuite toolkit]] which is focused around using the web platform as a
[[https://css-tricks.com/what-are-design-tokens/][design token]] development environment.

The core in particular is responsible for generating, assembling, and exporting token collections and dictionaries.

QuarkSuite Core is written in vanilla JavaScript (ES2017+).

* Features

+ Generate practical color palettes for apps and sites
  * Color adjustment and blending
  * Built-in colorblindness, contrast sensitivity, and illuminant simulations
  * Material-esque and artistic configurations
  * WCAG-based and colorimetic color accessibility filtering
+ Enforce compositional balance with modular scales
+ Export tokens to take with you
  * Graphic design palettes (GIMP/Inkscape, Sketch)
  * Stylesheet formats (CSS custom properties, Sass, Less, Stylus)
  * Data formats (JSON, YAML)
+ Interop with TailwindCSS and Style Dictionary
+ Scales with your needs
  * from basic generation and assembly
  * to advanced rules and processes
  * to modular configurations
+ Independent modules
  * =color.js= for color features
  * =content.js= for content features
  * =exporter.js= for exporting features
  * =workflow.js= for advanced features

* Packages

+ [[https://jsdelivr.com][JSDelivr]]: =https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/[module]=
+ [[https://statically.io][Statically]]: =https://cdn.statically.io/gh/quarksuite/core/v2.1.0/[module]=
+ [[https://nest.land][Nest.land]]: =https://x.nest.land/quarksuite:core@2.1.0/[module]=
+ [[https://nodejs.org/][Node]] (NPM): =@quarksuite-two/core@2.1.0=
  + [[https://skypack.dev][Skypack]]: =https://cdn.skypack.dev/@quarksuite-two/core@2.1.0/[module]=
  + [[https://esm.sh][esm.sh]]: =https://esm.sh/@quarksuite-two/core@2.1.0/[module]=

In general, prefer the JSDelivr/Statically CDN for availability, features, and speed. The Nest.land package is a
failsafe for the repo disappearing or GitHub going down.

* Getting Started

** Live

If you want to try the library live, load up your console in a modern browser or run a =deno repl= and copy/paste:

#+BEGIN_SRC js
const {
  color,
  content,
  exporter,
  workflow
} = await import("https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/mod.js");
#+END_SRC

** Local Development

For local dev, it's recommended that you [[https://buildless.site/][use a buildless stack]] or work in the browser. QuarkSuite Core is designed
especially for that.

*** Browser/Deno

#+BEGIN_SRC js
import {
  color,
  content,
  exporter,
  workflow
} from "https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/mod.js";
#+END_SRC

*** Node

#+BEGIN_QUOTE
The now defunct =@quarksuite/core= was originally published on NPM, but I started the project over and unpublished the
package. The core is now in a state where I feel comfortable putting it back on NPM and exposing it to the wider
JavaScript ecosystem.
#+END_QUOTE

Ensure that =type: module= is set in your =package.json= so that Node is set for ES Modules. Then install the package.

#+BEGIN_SRC shell
npm install @quarksuite-two/core@2.1.0

# OR

yarn add @quarksuite-two/core@2.1.0
#+END_SRC

And then in your source files:

#+BEGIN_SRC js
import {
  color,
  content,
  exporter,
  workflow
} from "@quarksuite-two/core";
#+END_SRC

** Offline

Some engines (like QuickJS) support ES Modules but don't allow importing from URLs. For those, you can =curl/wget=
the modules you need directly.

#+BEGIN_SRC shell
curl -o [module] https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/[module]

# OR

wget https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/[module]
#+END_SRC

* Handbook

The [[https://github.com/quarksuite/core/blob/main/HANDBOOK.org][QuarkSuite Core Handbook]] introduces the features of the library from a user perspective. It's recommended you read
that next.

* API

The [[https://github.com/quarksuite/core/blob/main/API.org][QuarkSuite Core API]] is a technical overview of the library that also serves as documentation of its functions. It
may make more sense after reading the handbook, but some of you may get more out of reading it beforehand.

* Demos

I maintain a [[https://glitch.com/@quarksuite][growing collection of demos]] for QuarkSuite at various stages of development on Glitch. It includes rough
experiments, workflow samples, and practical usage examples, so be sure to check it out.

* Contributing

First, clone the repo to your local machine.

#+BEGIN_SRC shell
git clone https://github.com/quarksuite/core.git
#+END_SRC

Then make sure you [[https://github.com/quarksuite/core/blob/main/CONTRIBUTING.md][read the contribution guidelines]].

* Concept

QuarkSuite is a project built around the idea that authoring design tokens shouldn't be limited to a particular
environment or web stack. Many design tools build entire systems around themselves.

I wanted a workflow where I could open a blank page in a browser and start experimenting. I wanted tools with mobility
across environments. I wanted something that could grow seamlessly with my needs and allow me to automate my
requirements.

QuarkSuite aims to be that toolkit.

* Project Objectives

The following objectives guide the ongoing development of QuarkSuite.

** Availability

This project focuses on the overall availability of web technologies over the popularity of particular web stacks. Its
independent modules, relatively small size, and simple output all work to make the kit extremely portable.

** Adaptability

QuarkSuite Core aims to grow or shrink according to your needs. Its functional nature makes it straightforward to use
and extend as you see fit.

** Invisibility

This project is a means to an end: creating design tokens. I don't want its processes to call attention to themselves or
its design to get in your way. Use what you need and leave the rest.

** Interoperability

QuarkSuite Core is built to work with what already exists in the design token and UI theming space. It can be used on
its own, but its schema is nothing but a plain object.

This makes it particularly suited for integration with design tools that use their /own/ schemas (such as [[https://tailwindcss.com/][TailwindCSS]]
and [[https://amzn.github.io/style-dictionary/][Style Dictionary]]).

** Stability

Where applicable, this project aims to adhere as closely as possible to current and emerging web standards. From its use
of standard CSS color formats to its avoidance of bleeding-edge syntax.

* Credit

+ Color conversion math and formulas
  - hex, rgb, hsl: [[https://www.rapidtables.com/convert/color/index.html][RapidTables]]
  - cmyk, hwb: [[https://drafts.csswg.org/css-color-4/][W3C CSS Color Module 4]] (though, as of this writing, =device-cmyk()= is bumped to Color Module 5)
  - lab, lch: [[http://www.brucelindbloom.com/index.html?Math.html][Bruce Lindbloom]]
  - oklab, oklch: [[https://bottosson.github.io/posts/oklab/][Björn Ottosson]] (the creator of the Oklab color space)
+ Color perception simulators
  - color vision deficiency: [[https://github.com/DaltonLens/libDaltonLens][libDaltonLens]]
  - correlated color temperature (CCT): [[https://github.com/m-lima/temperagb][temperargb]]
+ Built-in datasets:
  - Named colors: [[https://www.w3.org/wiki/CSS/Properties/color/keywords][W3C Wiki]]
  - System font stacks: [[https://systemfontstack.com][System Font Stack]]
+ design token schema: [[https://styled-system.com/theme-specification/][Styled System]] and [[https://amzn.github.io/style-dictionary/#/tokens][Style Dictionary]]


CHANGELOG.org
#+TITLE: Changelog

All notable changes to this project will be documented in this file.

The format is based on [[https://keepachangelog.com/en/1.0.0/][Keep a Changelog]] and this project adheres to [[https://semver.org/spec/v2.0.0.html][Semantic Versioning]].

* Table of Contents :TOC@3:
- [[#v210---2022-07-04][v2.1.0 - 2022-07-04]]
  - [[#added][Added]]
  - [[#changed][Changed]]
- [[#v202---2022-06-18][v2.0.2 - 2022-06-18]]
  - [[#fixed][Fixed]]
- [[#v201---2022-06-17][v2.0.1 - 2022-06-17]]
  - [[#fixed-1][Fixed]]
- [[#v200---2022-06-17][v2.0.0 - 2022-06-17]]
  - [[#added-1][Added]]
    - [[#simplified-workflow][simplified workflow]]
    - [[#independent-modules][independent modules]]
    - [[#tightened-api][tightened API]]
    - [[#faster-performance][faster performance]]
    - [[#color-accessibility-enhancements][color accessibility enhancements]]
    - [[#improved-developer-experience][improved developer experience]]
- [[#v130---2022-02-25][v1.3.0 - 2022-02-25]]
  - [[#added-2][Added]]
  - [[#changed-1][Changed]]
- [[#v121---2022-02-24][v1.2.1 - 2022-02-24]]
  - [[#fixed-2][Fixed]]
- [[#v120---2022-02-21][v1.2.0 - 2022-02-21]]
  - [[#changed-2][Changed]]
  - [[#fixed-3][Fixed]]
- [[#v110---2022-02-16][v1.1.0 - 2022-02-16]]
  - [[#added-3][Added]]
  - [[#changed-3][Changed]]
- [[#v101---2021-12-17][v1.0.1 - 2021-12-17]]
  - [[#added-4][Added]]
  - [[#fixed-4][Fixed]]
- [[#v100---2021-12-16][v1.0.0 - 2021-12-16]]
  - [[#added-5][Added]]
    - [[#revamped-workflow][revamped workflow]]
    - [[#explicit-naming-conventions][explicit naming conventions]]
    - [[#overhauled-api][overhauled API]]
- [[#v021---2020-06-07][v0.2.1 - 2020-06-07]]
  - [[#fixed-5][Fixed]]
    - [[#color-adjustment][color adjustment]]
- [[#v020---2020-05-27][v0.2.0 - 2020-05-27]]
  - [[#added-6][Added]]
    - [[#color-conversion][color conversion]]
  - [[#changed-4][Changed]]
    - [[#color-adjustment-1][color adjustment]]
    - [[#color-mixture][color mixture]]
    - [[#color-schemes][color schemes]]
    - [[#color-palette-color-scale][+color palette+ color scale]]
- [[#v011---2020-04-29][v0.1.1 - 2020-04-29]]
  - [[#changed-5][Changed]]
    - [[#color-conversion-1][color conversion]]
- [[#v010---2021-04-13][v0.1.0 - 2021-04-13]]
  - [[#added-7][Added]]
    - [[#functional-utilities][functional utilities]]
    - [[#color-conversion-2][color conversion]]
    - [[#scheme-functions][scheme functions]]
    - [[#color-accessibility][color accessibility]]
  - [[#changed-6][Changed]]
    - [[#scheme-functions-1][scheme functions]]
    - [[#output-functions][output functions]]
  - [[#removed][Removed]]
    - [[#functional-utilities-1][functional utilities]]

* v2.1.0 - 2022-07-04

This minor version moves palette accessibility and color perception simulation out of =palette()= as separate palette
actions. This makes them explicit operations to run /after/ creating a palette collection.

#+BEGIN_SRC js
import {
  palette,
  a11y,
  perception
} from "https://cdn.jsdelivr.net/gh/quarksuite/core@2.1.0/color.js";

const main = palette({ configuration: "material", accents: true }, "dodgerblue");

const ui = a11y({ mode: "standard", rating: "AA", large: true }, main);
const text = a11y({ mode: "standard", rating: "AA" }, main);

const checkUiProtanopia = perception({ check: "vision", as: "protanopia" }, ui);
#+END_SRC

See the updated handbook and [[https://github.com/quarksuite/core/blob/main/API.org][API documentation]].

** Added

+ =a11y(settings, palette)=
+ =perception(settings, palette)=

** Changed

=palette()= action now exclusively handles palette generation. Color accessibility and perception simulators are handled
by the above new palette actions.

* v2.0.2 - 2022-06-18

** Fixed

+ CIELCH color parsing didn't /actually/ convert the hue to radians, so CIELCH colors would shift their hue when
  converting to themselves.

* v2.0.1 - 2022-06-17

** Fixed

+ The =severity= setting was inactive for illuminant simulation

* v2.0.0 - 2022-06-17

QuarkSuite Core v2 is another complete rewrite of the library architecture that focuses on simplifying and tightening
the workflow introduced in v1. It also includes a significant performance boost and improves the developer experience.

For example:

#+BEGIN_SRC js
const color = convert("rgb", "dodgerblue");

const paletteOpts = { configuration: "material", accents: true };
const a11yOpts = { mode: "standard", rating: "AA" };

const font = "Work Sans";

const ratio = 1.618;
const values = 4;

const tokens = {
  color: {
    ui: palette({ ...paletteOpts, a11y: { ...a11yOpts, large: true } }, color),
    text: palette({ ...paletteOpts, a11y: a11yOpts }, color)
  },
  text: {
    body: text({ system: "sans", weights: ["regular", "bold" ]}, font),
    heading: text({ system: "serif", weights: ["light", "black"]}, font),
    size: scale({ configuration: "bidirectional", ratio, values }, "1rem"),
    measure: scale({ configuration: "ranged", floor: 45, trunc: true, ratio, values }, "75ch"),
    leading: scale({ configuration: "ranged", floor: 1.25, ratio, values }, 1.5)
  },
  spacing: scale({ configuration: "bidirectional", ratio, values }, "1ex"),
  grid: grid({ rows: 3, ratio }, 5),
  lengths: {
    width: scale({ configuration: "ranged", floor: 10, ratio, values }, "100vw"),
    height: scale({ configuration: "ranged", floor: 10, ratio, values }, "100vh"),
    shortest: scale({ configuration: "ranged", floor: 10, ratio, values }, "100vmin"),
    longest: scale({ configuration: "ranged", floor: 10, ratio, values }, "100vmax"),
  }
};
#+END_SRC

You can [[https://github.com/quarksuite/core/blob/main/HANDBOOK.org][read the new handbook]] for a practical introduction and then [[https://github.com/quarksuite/core/blob/main/API.org][dive into the API]] for the technical details.

** Added

*** simplified workflow

Core v2 reworks the utilities -> configurations -> formulas -> bootstrappers layered workflow into a more compact yet
powerful one based around actions and emitters.


exporter.js
/// <reference types="./types/exporter.d.ts" />

/**
 * @typedef {Partial<{
 *   name: string;
 *   author: string;
 *   version: string;
 *   license: string;
 *   bump: string;
 *   metadata: {
 *     description?: string;
 *     comments?: string;
 *   }
 * }>} ProjectSettings
 *
 * @typedef {string | number} TokenValue
 *
 * @typedef {{base: TokenValue; [variant: string]: TokenValue | TokenSubcategory }} TokenSubcategory
 *
 * @typedef {{ [category: string]: TokenValue | TokenSubcategory | {} | TokenSchema }} TokenSchema
 *
 * @typedef {{
 *   project: ProjectSettings;
 *   [tokens: string]: TokenSchema;
 * }} Dictionary
 */

/**
 * @typedef {"css" | "scss" | "less" | "styl"} StylesheetFormat
 * @typedef {"json" | "yaml"} DataFormat
 * @typedef {"tailwindcss" | "style-dictionary"} SchemaFormat
 */

/**
 * An exporter that takes a complete token `dict` and prepares a file-ready
 * template string for a given stylesheet `format`.
 *
 * @param {StylesheetFormat} format - the target stylesheet format
 * @param {Dictionary} dict - the input token dictionary
 * @returns {string} file-ready stylesheet output
 *
 * @remarks
 * As a rule, exporter functions do *not* assume read/write access to your system.
 * The output of an exporter will either be a prepared file-ready template string
 * (or stringified JSON) or an object according to its return type.
 *
 * You can then write this data to a file using your environment's native
 * filesystem API or a filesystem library of your choice.
 *
 * @example
 * Token dictionary stylesheet export example
 *
 * ```js
 * import { palette } from "https://x.nest.land/quarksuite:core@2.0.0/color.js";
 * import { text, scale } from "https://x.nest.land/quarksuite:core@2.0.0/content.js";
 * import { stylesheet } from "https://x.nest.land/quarksuite:core@2.0.0/exporter.js";
 *
 * const swatch = "dodgerblue";
 *
 * ratio = 1.5;
 * values = 6;
 *
 * const dict = {
 *   project: {
 *     name: "Example Stylesheet Variables",
 *     version: "0.1.0",
 *     author: "John Q. Public",
 *     license: "MIT",
 *   }, // required by every exporter
 *   esv: {
 *     color: palette({ configuration: "material" }, "dodgerblue"),
 *     text: {
 *       body: text({ system: "sans", weights: ["regular", "bold"] }, ""),
 *       size: scale({ configuration: "bidirectional", ratio, values }, "1rem"),
 *       measure: scale({ configuration: "ranged", floor: 45, trunc: true, ratio, values }, "75ch"),
 *       leading: scale({ configuration: "ranged", floor: 1.25, ratio, values }, 1.5),
 *       spacing: scale({ configuration: "bidirectional", ratio, values }, "1ex")
 *     }
 *   }
 * }
 *
 * export const css = stylesheet("css", dict);
 * export const scss = stylesheet("scss", dict);
 * export const less = stylesheet("less", dict);
 * export const styl = stylesheet("styl", dict);
 * ```
 */
export function stylesheet(format, dict) {
  return variables(format, dict);
}

/**
 * An exporter that takes a complete token `dict` and prepares a file-ready
 * template string for a given data `format`.
 *
 * @param {DataFormat} format - the target data format
 * @param {Dictionary} dict - the input token dictionary
 * @returns {string} file-ready data output
 *
 * @remarks
 * As a rule, exporter functions do *not* assume read/write access to your system.
 * The output of an exporter will either be a prepared file-ready template string
 * (or stringified JSON) or an object according to its return type.
 *
 * You can then write this data to a file using your environment's native
 * filesystem API or a filesystem library of your choice.
 *
 * @example
 * Token dictionary data export example
 *
 * ```js
 * import { palette } from "https://x.nest.land/quarksuite:core@2.0.0/color.js";
 * import { text, scale } from "https://x.nest.land/quarksuite:core@2.0.0/content.js";
 * import { data } from "https://x.nest.land/quarksuite:core@2.0.0/exporter.js";
 *
 * const swatch = "dodgerblue";
 *
 * ratio = 1.5;
 * values = 6;
 *
 * const dict = {
 *   project: {
 *     name: "Example Stylesheet Variables",
 *     version: "0.1.0",
 *     author: "John Q. Public",
 *     license: "MIT",
 *   }, // required by every exporter
 *   esv: {
 *     color: palette({ configuration: "material" }, "dodgerblue"),
 *     text: {
 *       body: text({ system: "sans", weights: ["regular", "bold"] }, ""),
 *       size: scale({ configuration: "bidirectional", ratio, values }, "1rem"),
 *       measure: scale({ configuration: "ranged", floor: 45, trunc: true, ratio, values }, "75ch"),
 *       leading: scale({ configuration: "ranged", floor: 1.25, ratio, values }, 1.5),
 *       spacing: scale({ configuration: "bidirectional", ratio, values }, "1ex")
 *     }
 *   }
 * }
 *
 * export const json = data("json", dict);
 * export const yaml = data("yaml", dict);
 * ```
 */
export function data(format, dict) {
  const output = {
    json: json.bind(null),
    yaml: yaml.bind(null),
  };

  return output[format](dict);
}

/**
 * An exporter that takes a complete token `dict` and translates the data
 * to a given `format` schema.
 *
 * @param {SchemaFormat} format - the target schema
 * @param {Dictionary} dict - the input token dictionary
 * @returns {object} the output data
 *
 * @remarks
 * As a rule, exporter functions do *not* assume read/write access to your system.
 * The output of an exporter will either be a prepared file-ready template string
 * (or stringified JSON) or an object according to its return type.
 *
 * You can then write this data to a file using your environment's native
 * filesystem API or a filesystem library of your choice.
 *
 * @example
 * Token dictionary schema translation example
 *
 * ```js
 * import { palette } from "https://x.nest.land/quarksuite:core@2.0.0/color.js";
 * import { text, scale } from "https://x.nest.land/quarksuite:core@2.0.0/content.js";
 * import { schema } from "https://x.nest.land/quarksuite:core@2.0.0/exporter.js";
 *
 * const swatch = "dodgerblue";
 *
 * ratio = 1.5;
 * values = 6;
 *
 * const dict = {
 *   project: {
 *     name: "Example Stylesheet Variables",
 *     version: "0.1.0",
 *     author: "John Q. Public",
 *     license: "MIT",
 *   }, // required by every exporter
 *   esv: {
 *     color: palette({ configuration: "material" }, "dodgerblue"),
 *     text: {
 *       body: text({ system: "sans", weights: ["regular", "bold"] }, ""),
 *       size: scale({ configuration: "bidirectional", ratio, values }, "1rem"),
 *       measure: scale({ configuration: "ranged", floor: 45, trunc: true, ratio, values }, "75ch"),
 *       leading: scale({ configuration: "ranged", floor: 1.25, ratio, values }, 1.5),
 *       spacing: scale({ configuration: "bidirectional", ratio, values }, "1ex")
 *     }
 *   }
 * }
 *
 * export const theme = schema("tailwindcss", dict);
 * export const tokens = schema("style-dictionary", dict);
 * ```
 */
export function schema(format, dict) {
  const output = {
    tailwindcss: tailwindcss.bind(null),
    "style-dictionary": styledictionary.bind(null),
  };

  return output[format](dict);
}

function timestamp() {
  const TIMESTAMP = new Date(Date.now());
  return `Updated on ${TIMESTAMP.toLocaleDateString()} at ${TIMESTAMP.toLocaleTimeString()}`;
}

function info(
  {
    commentDelim: [OPEN, DELIM, CLOSE] = ["\n  /**", "   * ", "\n   **/\n\n"],
    str = "",
  },
  metadata,
) {
  return str.concat(
    OPEN,
    Object.entries(metadata).reduce((str, [key, value]) => {
      const lines = value.split("\n");

      if (lines.length > 1) {
        return str
          .concat(
            "\n",
            DELIM,
            key.toUpperCase(),
            ":",
            "\n",
            DELIM.trimEnd(),
            lines.join(`\n${DELIM}`),
          )
          .trimEnd();
      }

      return str.concat("\n", DELIM, key.toUpperCase(), ": ", lines);
    }, ""),
    CLOSE,
  );
}

function bumpVersion(project) {
  let [major, minor, patch, pre] = Array.from(
    project.version.split(/[.-]/g),
  ).map((n) => parseFloat(n));

  function next(keyword) {
    const bumped = new Map([
      ["major", [major + 1, 0, 0]],
      ["minor", [major, minor + 1, 0]],
      ["patch", [major, minor, patch + 1]],
      ["pre", [major, minor, patch, pre + 1 || 0]],
      ["build", [major, minor, patch, pre, Date.now()]],
    ]).get(keyword);

    return bumped;
  }

  const releaseConditions = (release) =>
    Array.from(
      new Map([
        [release.length === 3, release.join(".")],
        [
          release.length === 4,
          [release.slice(0, 3).join("."), release[3]].join("-"),
        ],
        [
          release.length === 5,
          [
            release.slice(0, 3).join("."),
            [release[3] ?? 0, release[4]].join("+"),
          ].join("-"),
        ],
      ]),
    )
      .filter(([condition]) => condition)
      .flatMap(([, release]) => release)
      .toString();

  return { ...project, version: releaseConditions(next(project.bump)) };
}

function styleIdentifier(collected, current, delim) {
  if (current === "base") {
    return collected;
  }

  if (collected) {
    return [collected, current].join(delim);
  }

  return current;
}

function styleAssembler({
  padding = "",
  prefix = "--",
  assignment = ": ",
  suffix = ";",
  terminator = "\n",
}) {
  return function (str, key, value) {
    return str.concat(
      padding,
      prefix,
      key,
      assignment,
      value,
      suffix,
      terminator,
    );
  };
}

function styleFormatter(opts, node) {
  function assemble(head, node) {
    const { metadata, ...tokens } = node;

    return "".concat(
      (metadata && info(opts, metadata)) || "",
      Object.entries(tokens).reduce((str, [key, value]) => {
        const format = styleAssembler(opts);

        if (typeof value === "object") {
          return str.concat(assemble(styleIdentifier(head, key, "-"), value));
        }

        return format(str, styleIdentifier(head, key, "-"), value);
      }, ""),
    );
  }

  return "\n".concat(assemble("", node));
}

function styleHeader([DOC_OPEN, DOC_CLOSE] = ["\n/**", " **/\n"], project) {
  let {
    name = "Unknown",
    author = "Anonymous",
    version = "0.1.0",
    license = "Unlicense",
    bump = "manual",
    metadata: { description = "N/A", comments = "N/A" } = {},
  } = project;

  // Attach a dynamic property initializing the autorelease version
  const autobump = ["major", "minor", "patch", "pre", "build"].some(
    (keyword) => keyword === bump,
  );

  return "".concat(
    DOC_OPEN,
    `
 * Project: ${name} (v${autobump ? bumpVersion(project) : version})
 * Owned by: ${author}
 * License: ${license}
 * ${"=".repeat(80)}
${
      info(
        { commentDelim: [" *", " * ", ""] },
        {
          description,
          comments,
        },
      )
    }
 * ${"-".repeat(80)}
 * ${timestamp()}
`,
    DOC_CLOSE,
  );
}

function styleTokens(
  {
    metadata: [OPEN, DELIM, CLOSE] = ["\n  /**", "   * ", "\n   **/\n\n"],
    wrapper: [TOKENS_OPEN, TOKENS_CLOSE] = ["\n:root {", "\n}\n"],
    opts = { padding: "  " },
  },
  tokens,
) {
  return "".concat(
    TOKENS_OPEN,
    styleFormatter({ ...opts, commentDelim: [OPEN, DELIM, CLOSE] }, tokens),
    TOKENS_CLOSE,
  );
}

function style({ doc, ...opts } = {}, { project, ...tokens }) {
  return "".concat(styleHeader(doc, project), styleTokens(opts, tokens));
}

function variables(type, dict) {
  const format = {
    css: style.bind(null, {}),
    scss: style.bind(null, {
      doc: ["\n/*!", " */\n"],
      metadata: ["", "// ", "\n\n"],
      wrapper: ["", "\n"],
      opts: { padding: "", prefix: "$" },
    }),
    less: style.bind(null, {
      doc: ["\n/*", " */\n"],
      metadata: ["", "// ", "\n\n"],
      wrapper: ["", "\n"],
      opts: { padding: "", prefix: "@" },
    }),
    styl: style.bind(null, {
      doc: ["\n/*!", " */\n"],
      metadata: ["", "// ", "\n\n"],
      wrapper: ["", "\n"],
      opts: { padding: "", prefix: "", assignment: " = ", suffix: "" },
    }),
  };

  return format[type](dict);
}

function json(dict) {
  const { project, ...tokens } = dict;

  const { bump = "manual" } = project;

  // Check if bump matches an automation keyword
  const autobump = ["major", "minor", "patch", "pre", "build"].some(
    (keyword) => keyword === bump,
  );

  // Then bump the version
  autobump && bumpVersion(project);

  return JSON.stringify({ project, tokens }, null, 2);
}

function yamlValue(level, str, key, value) {
  const isMultiline = typeof value === "string" && value.split("\n").length > 1;
  if (isMultiline) {
    return str.concat(
      "".padStart(level),
      `${key}: |\n`,
      value
        .split("\n")
        .reduce((s, line) => s.concat("".padStart(level + 2), line, "\n"), ""),
    );
  }

  return str.concat("".padStart(level), key, ": ", value, "\n");
}

function yaml(dict) {
  const { project, ...tokens } = dict;
  const { bump = "manual" } = project;

  // Check if bump matches an automation keyword
  const autobump = ["major", "minor", "patch", "pre", "build"].some(
    (keyword) => keyword === bump,
  );

  // Then bump the version
  autobump && bumpVersion(project);

  // Recursively assemble the data tree
  const assemble = (level, tree) =>
    Object.entries(tree).reduce((str, [key, data]) => {
      if (typeof data === "string" || typeof data === "number") {
        return yamlValue(level, str, key, data);
      }
      return str.concat(
        "".padStart(level),
        key,
        ":\n",
        assemble(level + 2, data),
      );
    }, "");

  return `
# ${timestamp()}
${
    Object.entries({ project, tokens })
      .reduce((str, [key, data]) => {
        if (typeof data === "string" || typeof data === "number") {
          return yamlValue(0, str, key, data);
        }
        return str.concat("\n", key, ":\n", assemble(2, data));
      }, "")
      .trimEnd()
  }
`;
}

function tailwindcss(dict) {
  const { project, ...tokens } = dict;

  const assemble = (node) =>
    Object.entries(node).reduce((acc, [key, data]) => {
      if (key === "base") return { ...acc, DEFAULT: data };

      // Skip past any metadata
      if (key === "metadata") return { ...acc };

      if (typeof data === "object") {
        return { ...acc, [key]: assemble(data) };
      }

      return { ...acc, [key]: data };
    }, {});

  return project && assemble(tokens);
}

function styledictionary(dict) {
  const { project, ...tokens } = dict;

  const assemble = (node) =>
    Object.entries(node).reduce((acc, [key, data]) => {
      // Skip past any metadata
      if (key === "metadata") return { ...acc };

      if (typeof data === "object") {
        return { ...acc, [key]: assemble(data) };
      }

      return { ...acc, [key]: { value: data } };
    }, {});

  return project && assemble(tokens);
}
